crates/astria-core/src/protocol/transaction/v1alpha1/action_group/mod.rs

```diff
@@ -0,0 +1,201 @@
+}
+
+macro_rules! impl_belong_to_group {
+    ($(($act:ty, $group:expr)),*$(,)?) => {
+        $(
+            impl BelongsToGroup for $act {
+                const GROUP: ActionGroup = $group;
+            }
+        )*
+    }
+}
+
+impl_belong_to_group!(
+    (SequenceAction, ActionGroup::General),
+    (TransferAction, ActionGroup::General),
+    (ValidatorUpdate, ActionGroup::General),
+    (SudoAddressChangeAction, ActionGroup::UnbundleableSudo),
+    (IbcRelayerChangeAction, ActionGroup::Sudo),
+    (Ics20Withdrawal, ActionGroup::General),
+    (InitBridgeAccountAction, ActionGroup::UnbundleableGeneral),
+    (BridgeLockAction, ActionGroup::General),
+    (BridgeUnlockAction, ActionGroup::General),
+    (BridgeSudoChangeAction, ActionGroup::UnbundleableGeneral),
+    (FeeChangeAction, ActionGroup::Sudo),
+    (FeeAssetChangeAction, ActionGroup::Sudo),
+    (IbcRelay, ActionGroup::General),
+    (IbcSudoChangeAction, ActionGroup::UnbundleableSudo),
+);
+
+impl Action {
+    const fn group(&self) -> ActionGroup {
+        match self {
+            Action::Sequence(_) => SequenceAction::GROUP,
+            Action::Transfer(_) => TransferAction::GROUP,
+            Action::ValidatorUpdate(_) => ValidatorUpdate::GROUP,
+            Action::SudoAddressChange(_) => SudoAddressChangeAction::GROUP,
+            Action::IbcRelayerChange(_) => IbcRelayerChangeAction::GROUP,
+            Action::Ics20Withdrawal(_) => Ics20Withdrawal::GROUP,
+            Action::InitBridgeAccount(_) => InitBridgeAccountAction::GROUP,
+            Action::BridgeLock(_) => BridgeLockAction::GROUP,
+            Action::BridgeUnlock(_) => BridgeUnlockAction::GROUP,
+            Action::BridgeSudoChange(_) => BridgeSudoChangeAction::GROUP,
+            Action::FeeChange(_) => FeeChangeAction::GROUP,
+            Action::FeeAssetChange(_) => FeeAssetChangeAction::GROUP,
+            Action::Ibc(_) => IbcRelay::GROUP,
+            Action::IbcSudoChange(_) => IbcSudoChangeAction::GROUP,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum ActionGroup {
+    General,
+    UnbundleableGeneral,
+    Sudo,
+    UnbundleableSudo,
+}
+
+impl ActionGroup {
+    pub(super) fn is_bundleable(self) -> bool {
+        matches!(self, ActionGroup::General | ActionGroup::Sudo)
+    }
+
+    pub(super) fn is_sudo(self) -> bool {
+        matches!(self, ActionGroup::Sudo)
+    }
+}
+
+impl fmt::Display for ActionGroup {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ActionGroup::General => write!(f, "general"),
+            ActionGroup::UnbundleableGeneral => write!(f, "unbundleable general"),
+            ActionGroup::Sudo => write!(f, "sudo"),
+            ActionGroup::UnbundleableSudo => write!(f, "unbundleable sudo"),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error(transparent)]
+pub struct Error(ErrorKind);
+
+impl Error {
+    #[must_use]
+    pub fn kind(&self) -> &ErrorKind {
```

Comment:
private?

Reply:
I didn't know how to access the error for a cleaner testing style that Fraser recommended without making public: https://github.com/astriaorg/astria/blob/3e79a1c7794fe1fd78dcc25ae72de8b8f8b399c7/crates/astria-sequencer/src/app/tests_breaking_changes.rs#L525

Reply:
I think these tests are misplaced.

If I see it correctly, you want to check if the bundling logic works correctly, right? The correct place to test this is where the bundling is implemented, i.e. here in `astria-core`. `astria-sequencer` as a downstream consumer should be able to rely on the implementation being correct.

From a quick glance I could not see if there is any logic that branches on the error kind and subsequently takes a different path.

Reply:
There really isn't, but that's why we have a review process and rely on authors to note breaking changes.

But it is implicitly understood that changing behavior of fundamental types in `astria-core` cause immediate breakage in `astria-sequencer`.

Copying tests and publicly exposing implementation deatils for what's effectively a unit test in other parts of the code base just adds extra churn without giving you extra safety.